### ISSUE TITLE
Create loss function

### ISSUE THREAD

**Issue by ChristianEschen** [apache/incubator-mxnet#5972]
Hello!

I am about to implement a new loos function. My data is quite imbalanced, and thus I would like to pay more attention to the class with few samples.

It is regarding binary classification in image object detection.
The cost function I would like to implement is 
loss = w_pos * y * log sigma + w_neg * (1-y) * log (1-sigma)
 
data=mx.sym.Variable('data')
label=mx.sym.Variable('label')

out=mx.symbol.Softmax(data=data,name='softmax') # I dont know if 0 or 1
ce = weightpos*label * mx.sym.log(out) + weightneg*(1 - label) * mx.sym.log(1 - out)
loss = mx.sym.MakeLoss(ce, normalization='batch')

It fails to compile and train:

lease use mx.sym.Variable(init=mx.init.*) to set initialization pattern' % name)
ValueError: Unknown initialization pattern for label. Default initialization is now limited to "weight", "bias", "gamma" (1.0), and "beta" (0.0).Please use mx.sym.Variable(init=mx.init.*) to set initialization pattern

How do I create custom loss function?

**Comment by jmacglashan** [comment]
The problem isn't with the creation of your loss function. The issue is with how you're initializing your network weights. Specifically, it seems it's trying to initialize the values for your label variable, and it's complaining because it doesn't know an initialization strategy for a variable named label. Indeed, you shouldn't be initializing it, because it's a data input, not a weight :p 

So the question is, why did this happen? My suspicion is that you're using Module and didn't tell it the right name of the label variable. (I would need to see your additional code to know) Consider that this is in the init method signature for Module:
```
 def __init__(self, symbol, data_names=('data',), label_names=('softmax_label',),
                 logger=logging, context=ctx.cpu(), work_load_list=None,
                 fixed_param_names=None, state_names=None):
```
Note the default argument ` label_names=('softmax_label',)`. If you didn't set this value, then that's your problem, because in your case, your label variable is named 'label', not 'softmax_label'. Because it doesn't know that 'label' is the name of your label, it think it's a value to optimize, tries to initialize its value, and then complains because there is no defined strategy for parameters with the naming pattern 'label'.

tl;dr: do this:
```
mod = mx.module.Module(sym, ..., label_names=('label',), ...)
```

**Comment by jmacglashan** [comment]
P.S. I highly recommend that your final symbol group the prediction value with the loss, otherwise you'll only be able to get the output of the loss and not what it predicts :p That is,

```
final_out = mx.sym.Group([mx.sym.BlockGrad(out), loss])
```

(The block grad is used because there is no top gradient for the prediction and you don't want the gradients to be wrt it)

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
